commands/dial_stdio.go

```diff
@@ -0,0 +1,90 @@
+package commands
+
+import (
+	"io"
+	"os"
+
+	"github.com/docker/buildx/build"
+	"github.com/docker/buildx/builder"
+	"github.com/docker/buildx/util/progress"
+	"github.com/docker/cli/cli/command"
+	"github.com/moby/buildkit/util/appcontext"
+	"github.com/moby/buildkit/util/progress/progressui"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"golang.org/x/sync/errgroup"
+)
+
+type stdioOptions struct {
+	builder string
+}
+
+func runDialStdio(dockerCli command.Cli, opts stdioOptions) error {
+	ctx := appcontext.Context()
+
+	contextPathHash, _ := os.Getwd()
+	b, err := builder.New(dockerCli,
+		builder.WithName(opts.builder),
+		builder.WithContextPathHash(contextPathHash),
+	)
+	if err != nil {
+		return err
+	}
+
+	if err = updateLastActivity(dockerCli, b.NodeGroup); err != nil {
+		return errors.Wrapf(err, "failed to update builder last activity time")
+	}
+	nodes, err := b.LoadNodes(ctx)
+	if err != nil {
+		return err
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+
+	printer, err := progress.NewPrinter(ctx, os.Stderr, progressui.AutoMode)
+	if err != nil {
+		return err
+	}
+
+	return progress.Wrap("[internal] Dialing builder: "+b.Name, printer.Write, func(sub progress.SubLogger) error {
+		conn, err := build.Dial(ctx, nodes, printer)
+		if err != nil {
+			return err
+		}
+		defer conn.Close()
+
+		go func() {
```

Comment:
Could we extract the logic for dial-stdio in buildkit into a reusable function so we could just call that here after getting the `conn`?

Ideally, we wouldn't end up with diverging implementations in buildctl/buildx (e.g. that would make it a pain to fix bugs in the future, etc).

Reply:
One implementation of what exactly?
the main part of this is just wiring up copy operations.
I'm not sure its worth sharing an implementation of that?

Reply:
Ah, I think I'm just confused by the differences in handling `CloseRead`/`CloseWrite` here versus in buildctl. Given that they seem to be doing the same thing, I'm not quite sure I understand why the code is so different.

Reply:
@AkihiroSuda I'm curious on this case where its wrapping with a nop-closer.https://github.com/moby/buildkit/blame/5ae9b23c40a926615b6c3cc1da3f5457d2f61fd4/cmd/buildctl/dialstdio.go#L35

I don't think we necessarily need the stream to handle a half-close here (hence why its implemented here), just that it s a little cleaner in terms of the connection semantics.

Other than that I think the implementation is pretty much the same, just this is taking advantage of errgroup and buildctl is basically doing its own hand-rolled equivalent.

Reply:
From my memory, tty stuff didn't work correctly without handling half-close there

Reply:
I was trying to test for using tty as stdin/stdout but this broke http2.
Not sure how to check for this.

Fundamentally though, the only real change as compared to buildctl would be to not fallback to `Close` on when `CloseRead` is not implemented.